# Lectura de ficheros easyMoney

## commercial_activity.csv
|Variable|Explicación|
|---|---|
|*active_customer*| Indicador de actividad del cliente en nuestra aplicación
|*entry_channel*| Canal de captación del cliente
|*entry_date*| Fecha en la que realizó la primera contratación a través de easyMoney
|*segment*| Segmento comercial del cliente
|*pk_cid*| Identificador de cliente
|*pk_partition*| Fecha de ingesta de los datos

## sociodemographic.csv
|Variable|Explicación|
|---|---|
|*age*| Edad del cliente
|*country_id*| País de residencia del cliente
|*deceased*| Fallecido. N/S
|*gender*| Sexo del cliente
|*region_code*| Provincia de residencia del cliente (para ES)
|*salary*| Ingresos brutos de la unidad familiar

## products.csv
|Variable|Explicación|
|---|---|
|*credit_card*| Tarjetas de crédito
|*debit_card*| Tarjetas de débito
|*em_account_p*| Cuenta easyMoney+
|*em_account_pp*| Cuenta easyMoney++
|*em_acount*| Cuenta easyMoney
|*emc_account*| Cuenta easyMoney Crypto
|*funds*| Fondos de inversión
|*loans*| Préstamos
|*long_term_deposit*| Depósitos a largo plazo
|*mortgage*| Mortgage
|*payroll*| Domiciliaciones
|*payroll_account*| Cuenta bonificada por domiciliaciones
|*pension_plan*| Plan de pensiones
|*securities*| Valores
|*short_term_deposit*| Short-term deposits



## import Libraries

In [1]:
import numpy as np 
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns


pd.options.display.float_format = '{:,.2f}'.format

### Funcion para crear tabal de informacion de df

In [47]:
def build_my_info_table(df):
    # Seleccionar las columnas numéricas del DataFrame
    numerical_columns = df.select_dtypes(include=[np.number])
    
    # Crear un DataFrame con información detallada sobre las columnas
    df_info = pd.DataFrame({
        # Nombre de la columna
        'column':   [col             for col in df.columns],
        # Tipo de datos de la columna
        'dtype':    [df[col].dtype   for col in df.columns],
        # Número de valores no nulos en la columna
        'count':    [df[col].count() for col in df.columns],
        
        # Estadísticas descriptivas para columnas numéricas
        'mean':     [round(df[col].mean(), 1)         if col in numerical_columns else '' for col in df.columns],
        'std':      [round(df[col].std(), 1)          if col in numerical_columns else '' for col in df.columns],
        'min':      [round(df[col].min(), 1)          if col in numerical_columns else '' for col in df.columns],
        '25%':      [round(df[col].quantile(0.25), 1) if col in numerical_columns else '' for col in df.columns],
        '50%':      [round(df[col].median(), 1)       if col in numerical_columns else '' for col in df.columns],
        '75%':      [round(df[col].quantile(0.75), 1) if col in numerical_columns else '' for col in df.columns],
        'max':      [round(df[col].max(), 1)          if col in numerical_columns else '' for col in df.columns],
        
        # Número de valores únicos en la columna
        'nunique':  [df[col].nunique()                                          for col in df.columns],
        
        # Lista de valores únicos si el número de valores únicos es menor que 30, de lo contrario, cadena vacía
        'unique':   [list(df[col].unique()) if df[col].nunique() < 30 else ''   for col in df.columns],
        
        # Porcentaje de cardinalidad (número de valores únicos)
        'cardinality %': [round(df[col].nunique() / df.shape[0] * 100, 1)      for col in df.columns],
        
        # Moda (valor más frecuente) en la columna
        'mode':     [df[col].mode()[0]                                          for col in df.columns],
        
        # Porcentaje de ocurrencias de la moda en la columna
        'mode %':   [round(df[col].value_counts().max() * 100 / df.shape[0], 1) for col in df.columns],
        
        # Número de valores nulos en la columna
        'null_count': [df[col].isnull().sum() for col in df.columns],
        
        # Porcentaje de valores nulos en la columna
        'null %':   [round(df[col].isnull().mean() * 100, 1)                    for col in df.columns],
    })
    return df_info


In [2]:
df_c = pd.read_csv(r"C:\Users\Usuario\Desktop\Nuclio\TFM\easyMoney_fintech\data\commercial_activity_df.csv")

In [3]:
df_p = pd.read_csv(r"C:\Users\Usuario\Desktop\Nuclio\TFM\easyMoney_fintech\data\products_df.csv")

In [4]:
df_s = pd.read_csv(r"C:\Users\Usuario\Desktop\Nuclio\TFM\easyMoney_fintech\data\sociodemographic_df.csv")

In [5]:
df_c

Unnamed: 0   pk_cid pk_partition  entry_date entry_channel  \
0                 0  1375586   2018-01-28  2018-01-12           KHL   
1                 1  1050611   2018-01-28  2015-08-10           KHE   
2                 2  1050612   2018-01-28  2015-08-10           KHE   
3                 3  1050613   2018-01-28  2015-08-10           KHD   
4                 4  1050614   2018-01-28  2015-08-10           KHE   
...             ...      ...          ...         ...           ...   
5962919    13647304  1166765   2019-05-28  2016-08-14           KHE   
5962920    13647305  1166764   2019-05-28  2016-08-14           KHE   
5962921    13647306  1166763   2019-05-28  2016-08-14           KHE   
5962922    13647307  1166789   2019-05-28  2016-08-14           KHE   
5962923    13647308  1550586   2019-05-28  2019-05-13           NaN   

         active_customer             segment  
0                   1.00   02 - PARTICULARES  
1                   0.00  03 - UNIVERSITARIO  
2                   0.00  03 - UNIVERSITARIO  
3                   0.00  03 - UNIVERSITARIO  
4                   1.00  03 - UNIVERSITARIO  
...                  ...                 ...  
5962919             0.00  03 - UNIVERSITARIO  
5962920             0.00  03 - UNIVERSITARIO  
5962921             1.00   02 - PARTICULARES  
5962922             0.00  03 - UNIVERSITARIO  
5962923             1.00                 NaN  

[5962924 rows x 7 columns]

In [6]:
df_p

Unnamed: 0   pk_cid pk_partition  short_term_deposit  loans  \
0                 0  1375586   2018-01-28                   0      0   
1                 1  1050611   2018-01-28                   0      0   
2                 2  1050612   2018-01-28                   0      0   
3                 3  1050613   2018-01-28                   1      0   
4                 4  1050614   2018-01-28                   0      0   
...             ...      ...          ...                 ...    ...   
5962919    13647304  1166765   2019-05-28                   0      0   
5962920    13647305  1166764   2019-05-28                   0      0   
5962921    13647306  1166763   2019-05-28                   0      0   
5962922    13647307  1166789   2019-05-28                   0      0   
5962923    13647308  1550586   2019-05-28                   0      0   

         mortgage  funds  securities  long_term_deposit  em_account_pp  \
0               0      0           0                  0              0   
1               0      0           0                  0              0   
2               0      0           0                  0              0   
3               0      0           0                  0              0   
4               0      0           0                  0              0   
...           ...    ...         ...                ...            ...   
5962919         0      0           0                  0              0   
5962920         0      0           0                  0              0   
5962921         0      0           0                  0              0   
5962922         0      0           0                  0              0   
5962923         0      0           0                  0              0   

         credit_card  payroll  pension_plan  payroll_account  emc_account  \
0                  0     0.00          0.00                0            0   
1                  0     0.00          0.00                0            0   
2                  0     0.00          0.00                0            0   
3                  0     0.00          0.00                0            0   
4                  0     0.00          0.00                0            0   
...              ...      ...           ...              ...          ...   
5962919            0     0.00          0.00                0            0   
5962920            0     0.00          0.00                0            0   
5962921            0     0.00          0.00                0            0   
5962922            0     0.00          0.00                0            0   
5962923            0     0.00          0.00                0            0   

         debit_card  em_account_p  em_acount  
0                 0             0          1  
1                 0             0          1  
2                 0             0          1  
3                 0             0          0  
4                 0             0          1  
...             ...           ...        ...  
5962919           0             0          1  
5962920           0             0          1  
5962921           0             0          1  
5962922           0             0          1  
5962923           0             0          1  

[5962924 rows x 18 columns]

In [7]:
df_s

Unnamed: 0   pk_cid pk_partition country_id  region_code gender  age  \
0                 0  1375586   2018-01-28         ES        29.00      H   35   
1                 1  1050611   2018-01-28         ES        13.00      V   23   
2                 2  1050612   2018-01-28         ES        13.00      V   23   
3                 3  1050613   2018-01-28         ES        50.00      H   22   
4                 4  1050614   2018-01-28         ES        50.00      V   23   
...             ...      ...          ...        ...          ...    ...  ...   
5962919    13647304  1166765   2019-05-28         ES        50.00      V   22   
5962920    13647305  1166764   2019-05-28         ES        26.00      V   23   
5962921    13647306  1166763   2019-05-28         ES        50.00      H   47   
5962922    13647307  1166789   2019-05-28         ES        50.00      H   22   
5962923    13647308  1550586   2019-05-28         ES        28.00      H   37   

        deceased     salary  
0              N  87,218.10  
1              N  35,548.74  
2              N 122,179.11  
3              N 119,775.54  
4              N        NaN  
...          ...        ...  
5962919        N  43,912.17  
5962920        N  23,334.99  
5962921        N        NaN  
5962922        N 199,592.82  
5962923        N        NaN  

[5962924 rows x 9 columns]

Como se puede comprobar, a la hora de cargar el csv, se carga una columna "Unnamed:0", de la que no tengo informacion sobre lo que es. Puedo deducir que es una columna que viene como indice en el csv original, en principio como no tiene ninguna informacion relevante y ademas es diferente al indice generado por pandas( parece que hay alguna filas que salta de numero sin razon aparente) decido eliminarla ya que como identificador de cliente tengo "pk_cid".

In [8]:
df_p.drop('Unnamed: 0', axis=1, inplace=True)
df_c.drop('Unnamed: 0', axis=1, inplace=True)
df_s.drop('Unnamed: 0', axis=1, inplace=True)

In [13]:
# Elimino espacios de los nombres de las columnas y los paso a minusculas
df_s.columns = df_s.columns.str.replace(' ', '').str.lower()
df_c.columns = df_c.columns.str.replace(' ', '').str.lower()
df_p.columns = df_p.columns.str.replace(' ', '').str.lower()

## Primer análisis de la tabla "commercial_activity_df.csv"

In [9]:
df_c

pk_cid pk_partition  entry_date entry_channel  active_customer  \
0        1375586   2018-01-28  2018-01-12           KHL             1.00   
1        1050611   2018-01-28  2015-08-10           KHE             0.00   
2        1050612   2018-01-28  2015-08-10           KHE             0.00   
3        1050613   2018-01-28  2015-08-10           KHD             0.00   
4        1050614   2018-01-28  2015-08-10           KHE             1.00   
...          ...          ...         ...           ...              ...   
5962919  1166765   2019-05-28  2016-08-14           KHE             0.00   
5962920  1166764   2019-05-28  2016-08-14           KHE             0.00   
5962921  1166763   2019-05-28  2016-08-14           KHE             1.00   
5962922  1166789   2019-05-28  2016-08-14           KHE             0.00   
5962923  1550586   2019-05-28  2019-05-13           NaN             1.00   

                    segment  
0         02 - PARTICULARES  
1        03 - UNIVERSITARIO  
2        03 - UNIVERSITARIO  
3        03 - UNIVERSITARIO  
4        03 - UNIVERSITARIO  
...                     ...  
5962919  03 - UNIVERSITARIO  
5962920  03 - UNIVERSITARIO  
5962921   02 - PARTICULARES  
5962922  03 - UNIVERSITARIO  
5962923                 NaN  

[5962924 rows x 6 columns]

In [48]:
# Crear tabla de info.
df_c_info = build_my_info_table(df_c)

# Aplicar el estilo de gradiente de color.
styled_df_c = df_c_info.style.background_gradient(cmap='Reds', axis=0)

# Agregar estilos para hacer la tabla scrolleable y fijar los encabezados.
styled_df_c = styled_df_c.set_table_styles(
    [{
        'selector': 'thead th',
        'props': [('position', 'sticky'), ('top', '0'), ('background-color', 'black'), ('z-index', '1')]
    }, {
        'selector': 'tbody',
        'props': [('display', 'block'), ('max-height', '400px'), ('overflow-y', 'scroll'), ('overflow-x', 'auto')]
    }, {
        'selector': 'thead, tbody tr',
        'props': [('display', 'table'), ('width', '100%'), ('table-layout', 'fixed')]
    }, {
        'selector': 'th, td',
        'props': [('word-wrap', 'break-word'), ('white-space', 'pre-wrap'), ('max-width', '150px')]
    }]
)

# Mostrar la tabla.
styled_df_c

In [14]:
df_c.shape

(5962924, 6)

In [13]:
df_c.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5962924 entries, 0 to 5962923
Data columns (total 6 columns):
 #   Column           Dtype  
---  ------           -----  
 0   pk_cid           int64  
 1   pk_partition     object 
 2   entry_date       object 
 3   entry_channel    object 
 4   active_customer  float64
 5   segment          object 
dtypes: float64(1), int64(1), object(4)
memory usage: 273.0+ MB


In [16]:
df_c["active_customer"].value_counts()

active_customer
0.00    3557573
1.00    2405351
Name: count, dtype: int64

Viendo los tipos de variables que asigno pandas veo varios tipos que voy a cambiar.

"pk_cid" es un identificador de cliente en principio unico, lo comprobare mas adelante, no necesito que sea tipo numerico lo pasare a object.

"pk_partition" es una fecha asi que la convertire a tipo fecha.

"entry_date" es una fecha asi que la convertire a tipo fecha.

"active_consumer" solo tiene 0 y 1 como valores asi que lo cambio a tipo int.

In [35]:
# Cambiar el tipo de dato de las columnas

df_c['pk_cid'] = df_c['pk_cid'].astype('object')
df_c['pk_partition'] = pd.to_datetime(df_c['pk_partition'])
df_c['entry_date'] = pd.to_datetime(df_c['entry_date'], format='mixed', errors='coerce')
df_c['active_customer'] = df_c['active_customer'].astype('int64')


In [38]:
df_c.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5962924 entries, 0 to 5962923
Data columns (total 6 columns):
 #   Column           Dtype         
---  ------           -----         
 0   pk_cid           object        
 1   pk_partition     datetime64[ns]
 2   entry_date       datetime64[ns]
 3   entry_channel    object        
 4   active_customer  int64         
 5   segment          object        
dtypes: datetime64[ns](2), int64(1), object(3)
memory usage: 273.0+ MB


In [25]:
df_c.isna().sum()

pk_cid                  0
pk_partition            0
entry_date           6413
entry_channel      133033
active_customer         0
segment            133944
dtype: int64

In [26]:
df_c['pk_partition'].value_counts()

pk_partition
2019-05-28    442995
2019-04-28    439627
2019-03-28    436183
2019-02-28    431727
2019-01-28    426875
2018-12-28    422481
2018-11-28    416387
2018-10-28    402300
2018-09-28    375323
2018-08-28    352922
2018-07-28    339339
2018-06-28    252104
2018-05-28    249926
2018-04-28    247463
2018-03-28    245258
2018-02-28    242521
2018-01-28    239493
Name: count, dtype: int64

Por lo que podemos comporbar en un primer vistazo es que cada ingesta de datos de datos, se incrementa el numero de visualizaciones. Y podemos suponer que en este caso es el numero de clientes el que se incrementa al mes, teniendo datos de 239493 clientes en enero de 2018, y 442995 en la ultima ingesta de datos en mayo de 2019. A continuacion comprobare las demas tablas para corroborar esta hipótesis.

In [53]:
df_c[df_c['pk_cid'] == 1375586]

pk_cid pk_partition entry_date entry_channel  active_customer  \
0        1375586   2018-01-28 2018-01-12           KHL                1   
346322   1375586   2018-02-28 2018-01-12           KHL                1   
592407   1375586   2018-03-28 2018-01-12           KHL                1   
831104   1375586   2018-04-28 2018-01-12           KHL                1   
1108457  1375586   2018-05-28 2018-01-12           KHL                1   
1357280  1375586   2018-06-28 2018-01-12           KHL                1   
1791360  1375586   2018-07-28 2018-01-12           KHL                1   
1988936  1375586   2018-08-28 2018-01-12           KHL                0   
2487565  1375586   2018-09-28 2018-01-12           KHL                0   
2753984  1375586   2018-10-28 2018-01-12           KHL                0   
3156714  1375586   2018-11-28 2018-01-12           KHL                0   
3719100  1375586   2018-12-28 2018-01-12           KHL                0   
3950225  1375586   2019-01-28 2018-01-12           KHL                0   
4436665  1375586   2019-02-28 2018-01-12           KHL                0   
4840281  1375586   2019-03-28 2018-01-12           KHL                1   
5336125  1375586   2019-04-28 2018-01-12           KHL                0   
5569690  1375586   2019-05-28 2018-01-12           KHL                1   

                   segment  
0        02 - PARTICULARES  
346322   02 - PARTICULARES  
592407   02 - PARTICULARES  
831104   02 - PARTICULARES  
1108457  02 - PARTICULARES  
1357280  02 - PARTICULARES  
1791360  02 - PARTICULARES  
1988936  02 - PARTICULARES  
2487565  02 - PARTICULARES  
2753984  02 - PARTICULARES  
3156714  02 - PARTICULARES  
3719100  02 - PARTICULARES  
3950225  02 - PARTICULARES  
4436665  02 - PARTICULARES  
4840281  02 - PARTICULARES  
5336125  02 - PARTICULARES  
5569690  02 - PARTICULARES

Por lo que podemos comprobar en esta agrupacion con la info de un solo cliente vemos que cada mes se carga la info de dicho cliente con su informacion de ese mes.

## Analisis de la tabla "products_df.csv"

In [27]:
df_p.head()

pk_cid pk_partition  short_term_deposit  loans  mortgage  funds  \
0  1375586   2018-01-28                   0      0         0      0   
1  1050611   2018-01-28                   0      0         0      0   
2  1050612   2018-01-28                   0      0         0      0   
3  1050613   2018-01-28                   1      0         0      0   
4  1050614   2018-01-28                   0      0         0      0   

   securities  long_term_deposit  em_account_pp  credit_card  payroll  \
0           0                  0              0            0     0.00   
1           0                  0              0            0     0.00   
2           0                  0              0            0     0.00   
3           0                  0              0            0     0.00   
4           0                  0              0            0     0.00   

   pension_plan  payroll_account  emc_account  debit_card  em_account_p  \
0          0.00                0            0           0             0   
1          0.00                0            0           0             0   
2          0.00                0            0           0             0   
3          0.00                0            0           0             0   
4          0.00                0            0           0             0   

   em_acount  
0          1  
1          1  
2          1  
3          0  
4          1

In [49]:
# Crear tabla de info.
df_p_info = build_my_info_table(df_p)

# Aplicar el estilo de gradiente de color.
styled_df_p = df_p_info.style.background_gradient(cmap='Reds', axis=0)

# Agregar estilos para hacer la tabla scrolleable y fijar los encabezados.
styled_df_p = styled_df_p.set_table_styles(
    [{
        'selector': 'thead th',
        'props': [('position', 'sticky'), ('top', '0'), ('background-color', 'black'), ('z-index', '1')]
    }, {
        'selector': 'tbody',
        'props': [('display', 'block'), ('max-height', '400px'), ('overflow-y', 'scroll'), ('overflow-x', 'auto')]
    }, {
        'selector': 'thead, tbody tr',
        'props': [('display', 'table'), ('width', '100%'), ('table-layout', 'fixed')]
    }, {
        'selector': 'th, td',
        'props': [('word-wrap', 'break-word'), ('white-space', 'pre-wrap'), ('max-width', '150px')]
    }]
)

# Mostrar la tabla.
styled_df_p

Con un vistazo rapido de la tabla de informacion de cada columna, compruebo que toda las columnas son binarias menos el id y la fecha y solo dos tienen valores nulos, que mas adelante analizare para ver como tratarlos.

In [29]:
df_p['payroll'].value_counts()

payroll
0.00    5757076
1.00     205787
Name: count, dtype: int64

In [31]:
df_p['pension_plan'].value_counts()

pension_plan
0.00    5745061
1.00     217802
Name: count, dtype: int64

Tras comprobar que las variables con tipo float solo contienen valores de 1 y 0 cambio el tipo de variable a int.

In [42]:
df_p['payroll'] = df_p['payroll'].astype('Int64')
df_p['pension_plan'] = df_p['pension_plan'].astype('Int64')

In [43]:
df_p.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5962924 entries, 0 to 5962923
Data columns (total 17 columns):
 #   Column              Dtype 
---  ------              ----- 
 0   pk_cid              int64 
 1   pk_partition        object
 2   short_term_deposit  int64 
 3   loans               int64 
 4   mortgage            int64 
 5   funds               int64 
 6   securities          int64 
 7   long_term_deposit   int64 
 8   em_account_pp       int64 
 9   credit_card         int64 
 10  payroll             Int64 
 11  pension_plan        Int64 
 12  payroll_account     int64 
 13  emc_account         int64 
 14  debit_card          int64 
 15  em_account_p        int64 
 16  em_acount           int64 
dtypes: Int64(2), int64(14), object(1)
memory usage: 784.8+ MB


In [50]:
df_p['pk_partition'].value_counts()  


pk_partition
2019-05-28    442995
2019-04-28    439627
2019-03-28    436183
2019-02-28    431727
2019-01-28    426875
2018-12-28    422481
2018-11-28    416387
2018-10-28    402300
2018-09-28    375323
2018-08-28    352922
2018-07-28    339339
2018-06-28    252104
2018-05-28    249926
2018-04-28    247463
2018-03-28    245258
2018-02-28    242521
2018-01-28    239493
Name: count, dtype: int64

In [51]:
df_p[df_p['pk_cid'] == 1375586]

pk_cid pk_partition  short_term_deposit  loans  mortgage  funds  \
0        1375586   2018-01-28                   0      0         0      0   
346322   1375586   2018-02-28                   0      0         0      0   
592407   1375586   2018-03-28                   0      0         0      0   
831104   1375586   2018-04-28                   0      0         0      0   
1108457  1375586   2018-05-28                   0      0         0      0   
1357280  1375586   2018-06-28                   0      0         0      0   
1791360  1375586   2018-07-28                   0      0         0      0   
1988936  1375586   2018-08-28                   0      0         0      0   
2487565  1375586   2018-09-28                   0      0         0      0   
2753984  1375586   2018-10-28                   0      0         0      0   
3156714  1375586   2018-11-28                   0      0         0      0   
3719100  1375586   2018-12-28                   0      0         0      0   
3950225  1375586   2019-01-28                   0      0         0      0   
4436665  1375586   2019-02-28                   0      0         0      0   
4840281  1375586   2019-03-28                   0      0         0      0   
5336125  1375586   2019-04-28                   0      0         0      0   
5569690  1375586   2019-05-28                   0      0         0      0   

         securities  long_term_deposit  em_account_pp  credit_card  payroll  \
0                 0                  0              0            0        0   
346322            0                  0              0            0        0   
592407            0                  0              0            0        0   
831104            0                  0              0            0        0   
1108457           0                  0              0            0        0   
1357280           0                  0              0            0        0   
1791360           0                  0              0            0        0   
1988936           0                  0              0            0        0   
2487565           0                  0              0            0        0   
2753984           0                  0              0            0        0   
3156714           0                  0              0            0        0   
3719100           0                  0              0            0        0   
3950225           0                  0              0            0        0   
4436665           0                  0              0            0        0   
4840281           0                  0              0            0        0   
5336125           0                  0              0            0        0   
5569690           0                  0              0            0        0   

         pension_plan  payroll_account  emc_account  debit_card  em_account_p  \
0                   0                0            0           0             0   
346322              0                0            0           0             0   
592407              0                0            0           0             0   
831104              0                0            0           0             0   
1108457             0                0            0           1             0   
1357280             0                0            0           1             0   
1791360             0                0            0           1             0   
1988936             0                0            0           0             0   
2487565             0                0            0           0             0   
2753984             0                0            0           0             0   
3156714             0                0            0           0             0   
3719100             0                0            0           0             0   
3950225             0                0            0           0             0   
4436665             0                0            0           0   

Tambien compruebo, los datos del mismo cliente y veo que los indicen coinciden, tambien viendo los valores, deduzco que cunado hay un 1 en el producto ese cliente a usado y ha contratado el porducto el mes de la ingesta de datos 

### Sociodemographic

In [54]:
df_s.info(verbose=True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5962924 entries, 0 to 5962923
Data columns (total 8 columns):
 #   Column        Dtype  
---  ------        -----  
 0   pk_cid        int64  
 1   pk_partition  object 
 2   country_id    object 
 3   region_code   float64
 4   gender        object 
 5   age           int64  
 6   deceased      object 
 7   salary        float64
dtypes: float64(2), int64(2), object(4)
memory usage: 363.9+ MB


In [55]:
# Crear tabla de info.
df_s_info = build_my_info_table(df_s)

# Aplicar el estilo de gradiente de color.
styled_df_s = df_s_info.style.background_gradient(cmap='Reds', axis=0)

# Agregar estilos para hacer la tabla scrolleable y fijar los encabezados.
styled_df_s = styled_df_s.set_table_styles(
    [{
        'selector': 'thead th',
        'props': [('position', 'sticky'), ('top', '0'), ('background-color', 'black'), ('z-index', '1')]
    }, {
        'selector': 'tbody',
        'props': [('display', 'block'), ('max-height', '400px'), ('overflow-y', 'scroll'), ('overflow-x', 'auto')]
    }, {
        'selector': 'thead, tbody tr',
        'props': [('display', 'table'), ('width', '100%'), ('table-layout', 'fixed')]
    }, {
        'selector': 'th, td',
        'props': [('word-wrap', 'break-word'), ('white-space', 'pre-wrap'), ('max-width', '150px')]
    }]
)

# Mostrar la tabla.
styled_df_s

In [56]:
df_s[df_s['pk_cid'] == 1375586]

pk_cid pk_partition country_id  region_code gender  age deceased  \
0        1375586   2018-01-28         ES        29.00      H   35        N   
346322   1375586   2018-02-28         ES        29.00      H   35        N   
592407   1375586   2018-03-28         ES        29.00      H   35        N   
831104   1375586   2018-04-28         ES        29.00      H   35        N   
1108457  1375586   2018-05-28         ES        29.00      H   35        N   
1357280  1375586   2018-06-28         ES        29.00      H   35        N   
1791360  1375586   2018-07-28         ES        29.00      H   35        N   
1988936  1375586   2018-08-28         ES        29.00      H   36        N   
2487565  1375586   2018-09-28         ES        29.00      H   36        N   
2753984  1375586   2018-10-28         ES        29.00      H   36        N   
3156714  1375586   2018-11-28         ES        29.00      H   36        N   
3719100  1375586   2018-12-28         ES        29.00      H   36        N   
3950225  1375586   2019-01-28         ES        29.00      H   36        N   
4436665  1375586   2019-02-28         ES        29.00      H   36        N   
4840281  1375586   2019-03-28         ES        29.00      H   36        N   
5336125  1375586   2019-04-28         ES        29.00      H   36        N   
5569690  1375586   2019-05-28         ES        29.00      H   36        N   

           salary  
0       87,218.10  
346322  87,218.10  
592407  87,218.10  
831104  87,218.10  
1108457 87,218.10  
1357280 87,218.10  
1791360 87,218.10  
1988936 87,218.10  
2487565 87,218.10  
2753984 87,218.10  
3156714 87,218.10  
3719100 87,218.10  
3950225 87,218.10  
4436665 87,218.10  
4840281 87,218.10  
5336125 87,218.10  
5569690 87,218.10

Despues de analizar las tres tablas, llego a varias conclusiones:

- Las tres tablas tienen el mismo numero de filas y el mismo indice, asi que lo mas inteligente y para no tener varios dataframes, el siguiente paso sera unir los dataframes en uno solo.
- Hay columnas que tienen el mismo valor en todo sus observaciones lo mas logico sera eliminarlas, aunque es algo que tengo que hablar con el equipo.
- Los codigos de las regiones estan en numero y sera interesante crear una nueva columna o cambiarla para poder analizar los datos dependiendo de su provincia de una manera mas clara.
- Para hacer el primer vistazo general de como esta la empresa en este momento, y hacer un dashboard informativo creo que lo mas interesante seria utilizar solo los datos de la ultimna ingesta para hacer un primer analisis del estado actual de la empresa. 

In [ ]:
# Diccionario de mapeo de códigos de provincia a nombres de provincia

provincia_mapping = {
    1: "Álava",2: "Albacete",3: "Alicante",4: "Almería",5: "Ávila",6: "Badajoz",7: "Baleares, Islas",
    8: "Barcelona",9: "Burgos",10: "Cáceres",11: "Cádiz",12: "Castellón",13: "Ciudad Real",14: "Córdoba",15: "Coruña, A",
    16: "Cuenca",17: "Girona",18: "Granada",19: "Guadalajara",20: "Gipuzkoa",21: "Huelva",22: "Huesca",23: "Jaén",24: "León",
    25: "Lleida",26: "Rioja, La",27: "Lugo",28: "Madrid",29: "Málaga",30: "Murcia",31: "Navarra",32: "Ourense",33: "Asturias",
    34: "Palencia",35: "Palmas, Las",36: "Pontevedra",37: "Salamanca",38: "Santa Cruz de Tenerife",39: "Cantabria",40: "Segovia",
    41: "Sevilla",42: "Soria",43: "Tarragona",44: "Teruel",45: "Toledo",46: "Valencia",47: "Valladolid",48: "Bizkaia",49: "Zamora",
    50: "Zaragoza",51: "Ceuta",52: "Melilla"}
